### ISSUE TITLE
Question about the implementation of octree

### ISSUE THREAD

**Issue by fishfishson** [nv-tlabs/nglod#2]
Hi @tovacinni, thx for your excellent work! recently many sparse octree-based rendering papers are published e.g. [PlenOctrees](https://alexyu.net/plenoctrees/). However, in plenoctrees code for [building octree](https://pypi.org/project/svox/#files), they initialize the octree with a **dense buffer**, like 
```python
class N3Tree(nn.Module):
    """
    PyTorch :math:`N^3`-tree library with CUDA acceleration.
    By :math:`N^3`-tree we mean a 3D tree with branching factor N at each interior node,
    where :math:`N=2` is the familiar octree.

.. warning::
    `nn.Parameters` can change size, which
    makes current optimizers invalid. If any refine() or
    shrink_to_fit() call returns True,
    please re-make any optimizers
    """
    def __init__(self, N=2, data_dim=4, depth_limit=10,
            init_reserve=1, init_refine=0, geom_resize_fact=1.5,
            radius=0.5, center=[0.5, 0.5, 0.5],
            data_format="RGBA",
            extra_data=None,
            map_location="cpu"):
        """
        Construct N^3 Tree

        :param N: int branching factor N
        :param data_dim: int size of data stored at each leaf
        :param depth_limit: int maximum depth of tree to stop branching/refining
        :param init_reserve: int amount of nodes to reserve initially
        :param init_refine: int number of times to refine entire tree initially
        :param geom_resize_fact: float geometric resizing factor
        :param radius: float or list, 1/2 side length of cube (possibly in each dim)
        :param center: list center of space
        :param data_format: a string to indicate the data format
        :param extra_data: extra data to include with tree
        :param map_location: str device to put data

        """
        super().__init__()
        assert N >= 2
        assert depth_limit >= 0
        self.N : int = N
        self.data_dim : int = data_dim

        if init_refine > 0:
            for i in range(1, init_refine + 1):
                init_reserve += (N ** i) ** 3
        
        # Here N is the voxel size. 
        self.register_parameter("data", nn.Parameter(torch.zeros(init_reserve, N, N, N, data_dim, device=map_location)))
        self.register_buffer("child", torch.zeros(init_reserve, N, N, N, dtype=torch.int32, device=map_location))
```      
How's your implementation for octree? Will it still be sparse during the building and support large scale scene? 
Looking forward to your answer.

**Comment by fishfishson** [comment]
To be more clear, I want to use sparse voxel tree in both reconstruction and rendering, which requires a sparse storage structure during the whole training pipeline.

**Comment by tovacinni** [comment]
Thanks for your interest in our work. 

In the paper, at training time we similarly initialize with a dense buffer, and make it sparse at inference time. If you know or can predict which octree cells will be occupied, however, there's no reason to not be able to use sparsity to your advantage at training time. We plan to showcase some of these use-cases, but they probably won't be ready for the first release (which will hopefully come very soon). 

If you wanted to implement something like that yourself, though, all the code you'll need to implement that will be there in our first release.
